In [13]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm
from patsy import dmatrices

In [3]:
dta = sm.datasets.fair.load_pandas().data

In [4]:
dta.head()

rate_marriage   age  yrs_married  children  religious  educ  occupation  \
0            3.0  32.0          9.0       3.0        3.0  17.0         2.0   
1            3.0  27.0         13.0       3.0        1.0  14.0         3.0   
2            4.0  22.0          2.5       0.0        1.0  16.0         3.0   
3            4.0  37.0         16.5       4.0        3.0  16.0         5.0   
4            5.0  27.0          9.0       1.0        1.0  14.0         3.0   

   occupation_husb   affairs  
0              5.0  0.111111  
1              4.0  3.230769  
2              5.0  1.400000  
3              5.0  0.727273  
4              4.0  4.666666

In [5]:
dta.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6366 entries, 0 to 6365
Data columns (total 9 columns):
rate_marriage      6366 non-null float64
age                6366 non-null float64
yrs_married        6366 non-null float64
children           6366 non-null float64
religious          6366 non-null float64
educ               6366 non-null float64
occupation         6366 non-null float64
occupation_husb    6366 non-null float64
affairs            6366 non-null float64
dtypes: float64(9)
memory usage: 447.7 KB


In [6]:
dta['affairs'] = (dta['affairs']>0).astype(int)
dta.head()

rate_marriage   age  yrs_married  children  religious  educ  occupation  \
0            3.0  32.0          9.0       3.0        3.0  17.0         2.0   
1            3.0  27.0         13.0       3.0        1.0  14.0         3.0   
2            4.0  22.0          2.5       0.0        1.0  16.0         3.0   
3            4.0  37.0         16.5       4.0        3.0  16.0         5.0   
4            5.0  27.0          9.0       1.0        1.0  14.0         3.0   

   occupation_husb  affairs  
0              5.0        1  
1              4.0        1  
2              5.0        1  
3              5.0        1  
4              4.0        1

In [9]:
dta.groupby('affairs')['affairs'].count()

affairs
0    4313
1    2053
Name: affairs, dtype: int64

In [19]:
y, X = dmatrices('affairs ~ rate_marriage + age + yrs_married + children + religious \
                    + educ + C(occupation) + C(occupation_husb)',dta, return_type="dataframe")


In [20]:
X = X.rename(columns={'C(occupation)[T.2.0]':'occ_2',
                        'C(occupation)[T.3.0]':'occ_3',
                        'C(occupation)[T.4.0]':'occ_4',
                        'C(occupation)[T.5.0]':'occ_5',
                        'C(occupation)[T.6.0]':'occ_6',
                        'C(occupation_husb)[T.2.0]':'occ_husb_2',
                        'C(occupation_husb)[T.3.0]':'occ_husb_3',
                        'C(occupation_husb)[T.4.0]':'occ_husb_4',
                        'C(occupation_husb)[T.5.0]':'occ_husb_5',
                        'C(occupation_husb)[T.6.0]':'occ_husb_6'})


In [21]:
y = np.ravel(y)

In [22]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn import metrics

#### Split training and test data

In [48]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=34)

In [74]:
lr = LogisticRegression(C=0.001)
lr.fit(X_train, y_train)

LogisticRegression(C=0.001, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [75]:
# predict class labels for the test set
y_pred = lr.predict(X_test)
y_pred

array([0., 0., 0., ..., 0., 0., 0.])

In [76]:
lr.score(X_test,y_test)

0.706282722513089

We can also see the confusion matrix and a classification report with other metrics.

In [68]:
metrics.confusion_matrix(y_test, y_pred)

array([[1175,  126],
       [ 392,  217]])

In [69]:
print(metrics.classification_report(y_test, y_pred))

             precision    recall  f1-score   support

        0.0       0.75      0.90      0.82      1301
        1.0       0.63      0.36      0.46       609

avg / total       0.71      0.73      0.70      1910



#### Using Cross Validation

In [78]:
# evaluate the model using 10-fold cross-validation
scores = cross_val_score(LogisticRegression(), X, y, scoring='accuracy', cv=10)
print(scores)
print(np.mean(scores))

[0.72100313 0.70219436 0.73824451 0.70597484 0.70597484 0.72955975
 0.7327044  0.70440252 0.75157233 0.75      ]
0.7241630685514876


So, we got an average accuracy of 72.4% using cross validation